src/NodaTime/TimeInterval.cs

```diff
@@ -0,0 +1,284 @@
+        {
+            get
+            {
+                Preconditions.CheckState(start.HasValue, "TimeInterval extends to start of time");
+                return start.GetValueOrDefault();
+            }
+        }
+
+        /// <summary>
+        /// Returns <c>true</c> if this TimeInterval has a fixed start point, or <c>false</c> if it
+        /// extends to the start of time.
+        /// </summary>
+        /// <value><c>true</c> if this TimeInterval has a fixed start point, or <c>false</c> if it
+        /// extends to the start of time.</value>
+        public bool HasStart => start != null;
+
+        /// <summary>
+        /// Gets the end LocalTime - the exclusive upper bound of the TimeInterval.
+        /// </summary>
+        /// <value>The end <see cref="LocalTime"/>.</value>
+        /// <exception cref="InvalidOperationException">The TimeInterval extends to the end of time.</exception>
+        /// <seealso cref="HasEnd"/>
+        public LocalTime End
+        {
+            get
+            {
+                Preconditions.CheckState(end.HasValue, "TimeInterval extends to end of time");
+                return end.GetValueOrDefault();
+            }
+        }
+
+        /// <summary>
+        /// Returns <c>true</c> if this TimeInterval has a fixed end point, or <c>false</c> if it
+        /// extends to the end of time.
+        /// </summary>
+        /// <value><c>true</c> if this TimeInterval has a fixed end point, or <c>false</c> if it
+        /// extends to the end of time.</value>
+        public bool HasEnd => end.HasValue;
+
+
+        /// <summary>
+        /// Returns <c>true</c> if Start of TimeInterval is greater than End (for intervals like 23:00/1:00)
+        /// </summary>
+        public bool IsEndInNextDay { get; }
+
+        /// <summary>
+        /// Returns the Period of the TimeInterval.
+        /// </summary>
+        /// <remarks>
+        /// This will always be a non-negative Period, though it may be zero.
+        /// </remarks>
+        /// <value>The Period of the TimeInterval.</value>
+        /// <exception cref="InvalidOperationException">The TimeInterval extends to the start or end of time.</exception>
+        public Period Period
+        {
+            get
+            {
+                var duration = Duration;
+                return new Period(0, 0, 0, 0, 
+                    duration.Hours, duration.Minutes, duration.Seconds, duration.Milliseconds, 0, 
+                    duration.NanosecondOfDay % NodaConstants.NanosecondsPerMillisecond);
+            }
+        }
+
+        /// <summary>
+        /// Returns the Duration of the TimeInterval.
+        /// </summary>
+        /// <remarks>
+        /// This will always be a non-negative Duration, though it may be zero.
+        /// </remarks>
+        /// <value>The Duration of the TimeInterval.</value>
+        /// <exception cref="InvalidOperationException">The TimeInterval extends to the start or end of time.</exception>
```

Comment:
I'm not sure about this. It feels like a TimeInterval with no start and no end should be 24 hours long, and a single-end interval should have a duration based on the start/end of the day, e.g. 10pm- has a duration of 2 hours, and -8am has a duration of 8 hours.